client/allocrunner/taskrunner/envoybootstrap_hook_test.go

```diff
@@ -17,13 +17,222 @@ import (
 	"github.com/hashicorp/nomad/nomad/structs"
 	"github.com/stretchr/testify/require"
+	"golang.org/x/sys/unix"
 )
 
 var _ interfaces.TaskPrestartHook = (*envoyBootstrapHook)(nil)
 
+func writeTmp(t *testing.T, s string, fm os.FileMode) string {
+	dir, err := ioutil.TempDir("", "envoy-")
+	require.NoError(t, err)
+
+	fPath := filepath.Join(dir, sidsTokenFile)
+	err = ioutil.WriteFile(fPath, []byte(s), fm)
+	require.NoError(t, err)
+
+	return dir
+}
+
+func TestEnvoyBootstrapHook_maybeLoadSIToken(t *testing.T) {
+	t.Parallel()
+
+	if unix.Geteuid() == 0 {
+		t.Skip("test only works as non-root")
+	}
+
+	t.Run("file does not exist", func(t *testing.T) {
+		h := newEnvoyBootstrapHook(&envoyBootstrapHookConfig{logger: testlog.HCLogger(t)})
+		config, err := h.maybeLoadSIToken("task1", "/does/not/exist")
+		require.NoError(t, err) // absence of token is not an error
+		require.Equal(t, "", config)
+	})
+
+	t.Run("load token from file", func(t *testing.T) {
+		token := uuid.Generate()
+		f := writeTmp(t, token, 0440)
+		defer cleanupDir(t, f)
+
+		h := newEnvoyBootstrapHook(&envoyBootstrapHookConfig{logger: testlog.HCLogger(t)})
+		config, err := h.maybeLoadSIToken("task1", f)
+		require.NoError(t, err)
+		require.Equal(t, token, config)
+	})
+
+	t.Run("file is unreadable", func(t *testing.T) {
+		token := uuid.Generate()
+		f := writeTmp(t, token, 0200)
+		defer cleanupDir(t, f)
+
+		h := newEnvoyBootstrapHook(&envoyBootstrapHookConfig{logger: testlog.HCLogger(t)})
+		config, err := h.maybeLoadSIToken("task1", f)
+		require.Error(t, err)
+		require.False(t, os.IsNotExist(err))
+		require.Equal(t, "", config)
+	})
+}
+
+func TestEnvoyBootstrapHook_envoyBootstrapArgs(t *testing.T) {
+	t.Parallel()
+
+	t.Run("excluding SI token", func(t *testing.T) {
+		ebArgs := envoyBootstrapArgs{
+			sidecarFor:     "s1",
+			grpcAddr:       "1.1.1.1",
+			consulHTTPAddr: "2.2.2.2",
+		}
+		args := ebArgs.args()
+		require.Equal(t, []string{"connect", "envoy",
+			"-grpc-addr", "1.1.1.1",
+			"-http-addr", "2.2.2.2",
+			"-bootstrap",
+			"-sidecar-for", "s1",
+		}, args)
+	})
+
+	t.Run("including SI token", func(t *testing.T) {
+		token := uuid.Generate()
+		ebArgs := envoyBootstrapArgs{
+			sidecarFor:     "s1",
+			grpcAddr:       "1.1.1.1",
+			consulHTTPAddr: "2.2.2.2",
+			siToken:        token,
+		}
+		args := ebArgs.args()
+		require.Equal(t, []string{"connect", "envoy",
+			"-grpc-addr", "1.1.1.1",
+			"-http-addr", "2.2.2.2",
+			"-bootstrap",
+			"-sidecar-for", "s1",
+			"-token", token,
+		}, args)
+	})
+}
+
+// dig through envoy config to look for consul token
+type envoyConfig struct {
+	DynamicResources struct {
+		ADSConfig struct {
+			GRPCServices struct {
+				InitialMetadata []struct {
+					Key   string `json:"key"`
+					Value string `json:"value"`
+				} `json:"initial_metadata"`
+			} `json:"grpc_services"`
+		} `json:"ads_config"`
+	} `json:"dynamic_resources"`
+}
+
+func TestEnvoyBootstrapHook_with_SI_token(t *testing.T) {
```

Comment:
For larger tests especially I like a comment to summarize their assertions like:


```suggestion
// TestEnvoyBootstrapHook_with_SI_token asserts the bootstrap file written for
// Envoy contains a Consul SI token.
func TestEnvoyBootstrapHook_with_SI_token(t *testing.T) {
```

That way when somebody's code changes break it hopefully they have some idea of what assertions to preserve when they're fixing it.